samza-azure/src/main/java/org/apache/samza/AsyncSystemProducer.java

```diff
@@ -0,0 +1,182 @@
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import org.apache.samza.config.Config;
+import org.apache.samza.config.StreamConfig;
+import org.apache.samza.metrics.Counter;
+import org.apache.samza.metrics.MetricsRegistry;
+import org.apache.samza.system.OutgoingMessageEnvelope;
+import org.apache.samza.system.SystemProducer;
+import org.apache.samza.system.eventhub.metrics.SamzaHistogram;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * SystemProducer Helper that can be used to implement SystemProducer that provide async send APIs
+ * e.g. Kinesis, Event Hubs, etc..
+ * To implement an AsyncSystemProducer, you need to implement {@link AsyncSystemProducer#sendAsync}
+ */
+public abstract class AsyncSystemProducer implements SystemProducer {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AsyncSystemProducer.class.getName());
+
+  /**
+   * The constant CONFIG_STREAM_LIST.
+   */
+  public static final String CONFIG_STREAM_LIST = "systems.%s.stream.list";
+
+  private static final String SEND_ERRORS = "sendErrors";
+  private static final String SEND_CALLBACK_LATENCY = "sendCallbackLatency";
+  private static final String SEND_LATENCY = "sendLatency";
+
+  /**
+   * The constant AGGREGATE.
+   */
+  protected static final String AGGREGATE = "aggregate";
+
+  /**
+   * The Stream ids.
+   */
+  protected final List<String> streamIds;
+  /**
+   * The Physical to stream ids.
+   */
+  protected final Map<String, String> physicalToStreamIds;
+  /**
+   * The Registry.
+   */
+  protected final MetricsRegistry registry;
+
+  /**
+   * The Pending futures.
+   */
+  protected final Set<CompletableFuture<Void>> pendingFutures = ConcurrentHashMap.newKeySet();
+
+  private final AtomicReference<Throwable> sendExceptionOnCallback = new AtomicReference<>(null);
+
+  private static Counter aggSendErrors = null;
+  private static SamzaHistogram aggSendLatency = null;
+  private static SamzaHistogram aggSendCallbackLatency = null;
+
+  private final HashMap<String, SamzaHistogram> sendLatency = new HashMap<>();
+  private final HashMap<String, SamzaHistogram> sendCallbackLatency = new HashMap<>();
+  private final HashMap<String, Counter> sendErrors = new HashMap<>();
+
+
+
+  /**
+   * Instantiates a new Async system producer.
+   *
+   * @param systemName the system name
+   * @param config the config
+   * @param registry the registry
+   */
+  public AsyncSystemProducer(String systemName, Config config, MetricsRegistry registry) {
+    StreamConfig sconfig = new StreamConfig(config);
+    streamIds = config.getList(String.format(CONFIG_STREAM_LIST, systemName));
+    physicalToStreamIds =
+        streamIds.stream().collect(Collectors.toMap(sconfig::getPhysicalName, Function.identity()));
+    this.registry = registry;
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void send(String source, OutgoingMessageEnvelope envelope) {
```

Comment:
comment: If you decide to synchronize access to this pendingFutures, I suggest you add a synchronized modified to this method.

Reply:
As discussed, We are removing the pendingFutures on callback. As an added protection for EventHubSystemProducer we will clear the pendingFutures once the flush is called.